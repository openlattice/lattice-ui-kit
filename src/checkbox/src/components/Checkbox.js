// @flow
import React, { Component } from 'react';
import { CheckboxIndicator, CheckboxInput, CheckboxLabel } from './styled';

type Props = {
  disabled ? :boolean;
  onChange ? :(event :SyntheticInputEvent<HTMLInputElement>) => mixed;
  value ? :any;
  label ? :string;
  name ? :string;
  checked ? :boolean;
  defaultChecked ? :boolean;
  readOnly ? :boolean;
}

type State = {
  isChecked :boolean;
};

class Checkbox extends Component<Props, State> {

  static defaultProps = {
    checked: undefined,
    defaultChecked: false,
    disabled: false,
    label: undefined,
    name: undefined,
    onChange: undefined,
    readOnly: undefined,
    value: undefined,
  };

  constructor(props :Props) {
    super(props);

    const { checked, defaultChecked } = props;
    this.state = {
      isChecked: (checked !== undefined)
        ? checked
        // $FlowFixMe optional prop provided by defaultProps
        : defaultChecked,
    };
  }

  // Use prop 'checked' over state 'isChecked' if available
  getIsChecked = () => {
    const { checked } = this.props;
    const { isChecked } = this.state;
    return checked !== undefined ? checked : isChecked;
  }

  handleChange = (e :SyntheticInputEvent<HTMLInputElement>) => {
    const { disabled, onChange, readOnly } = this.props;
    e.persist();
    if (!(disabled || readOnly)) {
      this.setState({ isChecked: e.target.checked });
      if (onChange) {
        onChange(e);
      }
    }
  };

  render() {
    const {
      checked,
      defaultChecked,
      label,
      onChange,
      readOnly,
      ...rest
    } = this.props;
    const isChecked = this.getIsChecked();
    return (
      <CheckboxLabel readOnly={readOnly}>
        {label}
        <CheckboxInput
            checked={isChecked}
            onChange={this.handleChange}
            {...rest} />
        <CheckboxIndicator checked={isChecked} />
      </CheckboxLabel>
    );
  }
}

export default Checkbox;
