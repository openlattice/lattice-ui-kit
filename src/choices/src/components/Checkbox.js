// @flow
import { forwardRef } from 'react';

import {
  CheckboxButtonIndicator,
  CheckboxChipIndicator,
  CheckboxIndicator,
  CheckboxInput,
  ChoiceInnerWrapper,
  ChoiceLabel,
  ChoiceText,
  ChoiceWrapper,
} from './styled';

import { stopPropagation } from '../../../utils';

type Props = {|
  checked ?:boolean;
  defaultChecked ?:boolean;
  disabled ?:boolean;
  forwardedRef ?:any;
  id ?:string;
  label ?:string;
  mode ?:string;
  name ?:string;
  onBlur ?:(event :SyntheticFocusEvent<HTMLInputElement>) => void;
  onChange ?:(event :SyntheticInputEvent<HTMLInputElement>) => void;
  onFocus ?:(event :SyntheticFocusEvent<HTMLInputElement>) => void;
  readOnly ?:boolean;
  value ?:any;
|};

/* eslint-disable react/jsx-props-no-spreading */
const Checkbox = ({
  disabled,
  forwardedRef,
  id,
  label,
  mode,
  readOnly,
  ...rest
} :Props) => (
  <ChoiceLabel htmlFor={id} readOnly={readOnly} disabled={disabled}>
    <ChoiceWrapper mode={mode}>
      <ChoiceInnerWrapper>
        <CheckboxInput
            id={id}
            readOnly={readOnly}
            disabled={disabled || readOnly}
            {...rest}
            ref={forwardedRef} />
        {
          mode === 'button' && label
            && <CheckboxButtonIndicator onClick={stopPropagation}>{label}</CheckboxButtonIndicator>
        }
        {
          mode === 'chip' && label
            && <CheckboxChipIndicator onClick={stopPropagation}>{label}</CheckboxChipIndicator>
        }
        { mode !== 'button' && mode !== 'chip' && <CheckboxIndicator />}
      </ChoiceInnerWrapper>
    </ChoiceWrapper>
    {
      (mode !== 'button' && mode !== 'chip') && (
        <ChoiceText>
          {label}
        </ChoiceText>
      )
    }
  </ChoiceLabel>
);
/* eslint-enable */

Checkbox.defaultProps = {
  checked: undefined,
  defaultChecked: undefined,
  disabled: false,
  forwardedRef: undefined,
  id: undefined,
  label: undefined,
  mode: undefined,
  name: undefined,
  onBlur: undefined,
  onChange: undefined,
  onFocus: undefined,
  readOnly: undefined,
  value: undefined,
};

export default forwardRef<Props, HTMLInputElement>((props, ref) => (
  /* eslint-disable react/jsx-props-no-spreading */
  <Checkbox {...props} forwardedRef={ref} />
  /* eslint-enable */
));
