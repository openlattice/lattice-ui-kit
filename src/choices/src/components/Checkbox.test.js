import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';

import Checkbox from './Checkbox';
import {
  CheckboxButtonIndicator,
  CheckboxIndicator,
  CheckboxInput,
  CheckboxPillIndicator,
  ChoiceLabel
} from './styled';

describe('Checkbox', () => {

  test('render matches snapshot', () => {
    const wrapper = mount(<Checkbox />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('should pass through any props to the input', () => {
    const nativeProps = {
      disabled: true,
      name: 'test',
      required: true,
      form: 'test-form',
    };

    // eslint-disable-next-line react/jsx-props-no-spreading
    const wrapper = mount(<Checkbox {...nativeProps} />)
      .find('input')
      .props();
    expect(wrapper).toEqual(expect.objectContaining(nativeProps));
  });

  describe('Checkbox', () => {

    test('should be unchecked by default', () => {
      const wrapper = mount(<Checkbox />);
      expect(wrapper.find('input').prop('checked')).toEqual(undefined);
    });

    test('should render ChoiceLabel', () => {
      const wrapper = mount(<Checkbox label="test" />);
      expect(wrapper.find(ChoiceLabel).text()).toBe('test');
    });

    test('should render CheckboxIndicator', () => {
      const wrapper = mount(<Checkbox />);
      expect(wrapper.find(CheckboxIndicator).exists()).toBe(true);
    });

    test('should render CheckboxInput', () => {
      const wrapper = mount(<Checkbox />);
      expect(wrapper.find(CheckboxInput).exists()).toBe(true);
    });

  });

  describe('onChange', () => {

    test('should call onChange once', () => {
      const mockOnChange = jest.fn();
      const wrapper = mount(<Checkbox onChange={mockOnChange} />);

      wrapper.find(CheckboxInput).simulate('change', { target: { checked: true } });
      expect(mockOnChange).toHaveBeenCalledTimes(1);
    });

  });

  describe('mode', () => {

    test('should render CheckboxButtonIndicator when passed a label and mode="button"', () => {
      const wrapper = mount(<Checkbox label="Checkbox Button" mode="button" />);
      expect(wrapper.find(CheckboxButtonIndicator).exists()).toBe(true);
    });

    test('should render CheckboxPillIndicator when passed a label and mode="pill"', () => {
      const wrapper = mount(<Checkbox label="Checkbox Pill" mode="pill" />);
      expect(wrapper.find(CheckboxPillIndicator).exists()).toBe(true);
    });
  });

});
