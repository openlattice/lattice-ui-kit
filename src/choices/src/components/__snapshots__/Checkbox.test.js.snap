// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox render matches snapshot 1`] = `
.c4 {
  -webkit-transition: background-color 100ms ease-out,border-color 100ms ease-out,box-shadow 100ms ease-out,color 100ms ease-out;
  transition: background-color 100ms ease-out,border-color 100ms ease-out,box-shadow 100ms ease-out,color 100ms ease-out;
  background-color: #e6e6f7;
  border-radius: 3px;
  display: inline-block;
  height: 20px;
  width: 20px;
}

input:hover ~ .c4 {
  background-color: #8e929b;
}

input:focus ~ .c4 {
  box-shadow: #410ab5 0 0 0 2px;
}

input:checked ~ .c4,
input[readonly]:checked:disabled ~ .c4,
input[readonly]:checked:hover ~ .c4 {
  background-color: #8045ff;
}

input:checked:hover ~ .c4 {
  background-color: #6124e2;
}

input:disabled ~ .c4 {
  background-color: #e6e6f7;
  cursor: not-allowed;
}

input:checked:disabled ~ .c4 {
  cursor: not-allowed;
  background-color: #b6bbc7;
}

input[readonly] ~ .c4 {
  cursor: default;
  pointer-events: none;
}

input:not(:checked) ~ .c4:after,
input:checked ~ .c4:after {
  border-color: #ffffff;
  border-style: solid;
  border-width: 0 2px 2px 0;
  box-sizing: border-box;
  content: '';
  height: 10px;
  left: 7px;
  position: absolute;
  top: 4px;
  -webkit-transform: rotate(45deg) scale(1);
  -ms-transform: rotate(45deg) scale(1);
  transform: rotate(45deg) scale(1);
  -webkit-transition: -webkit-transform 200ms ease-out,opacity 200ms ease-out;
  -webkit-transition: transform 200ms ease-out,opacity 200ms ease-out;
  transition: transform 200ms ease-out,opacity 200ms ease-out;
  width: 6px;
}

input:not(:checked) ~ .c4:after {
  opacity: 0;
  -webkit-transform: rotate(45deg) scale(0);
  -ms-transform: rotate(45deg) scale(0);
  transform: rotate(45deg) scale(0);
}

input:checked ~ .c4:after {
  opacity: 1;
  -webkit-transform: rotate(45deg) scale(1);
  -ms-transform: rotate(45deg) scale(1);
  transform: rotate(45deg) scale(1);
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #555e6f;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 14px;
  min-height: 40px;
  pointer-events: auto;
  position: relative;
  vertical-align: middle;
}

.c5 {
  margin-left: 10px;
}

.c3 {
  height: 100%;
  left: 0;
  margin: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  vertical-align: middle;
  width: 100%;
  z-index: -1;
}

.c1 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  vertical-align: middle;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<Checkbox
  disabled={false}
>
  <ChoiceLabel
    disabled={false}
  >
    <StyledComponent
      disabled={false}
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "ChoiceLabel-sc-11kob9e-0",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "align-items:center;color:",
              [Function],
              ";display:inline-flex;font-size:14px;min-height:40px;pointer-events:",
              [Function],
              ";position:relative;vertical-align:middle;",
            ],
          },
          "displayName": "ChoiceLabel",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "ChoiceLabel-sc-11kob9e-0",
          "target": "label",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <label
        className="c0"
        disabled={false}
      >
        <ChoiceWrappers__ChoiceWrapper>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "ChoiceWrappers__ChoiceWrapper-kj63qq-0",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "align-items:stretch;display:inline-flex;justify-content:center;position:relative;vertical-align:middle;",
                    [Function],
                    ";",
                  ],
                },
                "displayName": "ChoiceWrappers__ChoiceWrapper",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "ChoiceWrappers__ChoiceWrapper-kj63qq-0",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <span
              className="c1"
            >
              <ChoiceWrappers__ChoiceInnerWrapper>
                <StyledComponent
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "ChoiceWrappers__ChoiceInnerWrapper-kj63qq-1",
                        "isStatic": false,
                        "lastClassName": "c2",
                        "rules": Array [
                          "display:flex;flex:1;",
                        ],
                      },
                      "displayName": "ChoiceWrappers__ChoiceInnerWrapper",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "ChoiceWrappers__ChoiceInnerWrapper-kj63qq-1",
                      "target": "span",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                >
                  <span
                    className="c2"
                  >
                    <ChoiceInputs__CheckboxInput>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [
                              [Function],
                            ],
                            "componentStyle": ComponentStyle {
                              "componentId": "ChoiceInputs__CheckboxInput-sc-1fkg6yx-1",
                              "isStatic": false,
                              "lastClassName": "c3",
                              "rules": Array [
                                "height:100%;left:0;margin:0;opacity:0;position:absolute;top:0;vertical-align:middle;width:100%;z-index:-1;",
                                ";",
                              ],
                            },
                            "displayName": "ChoiceInputs__CheckboxInput",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "ChoiceInputs__CheckboxInput-sc-1fkg6yx-1",
                            "target": "input",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <input
                          className="c3"
                          type="checkbox"
                        />
                      </StyledComponent>
                    </ChoiceInputs__CheckboxInput>
                    <CheckboxIndicator>
                      <StyledComponent
                        forwardedComponent={
                          Object {
                            "$$typeof": Symbol(react.forward_ref),
                            "attrs": Array [],
                            "componentStyle": ComponentStyle {
                              "componentId": "CheckboxIndicator-sc-16t4mhx-0",
                              "isStatic": false,
                              "lastClassName": "c4",
                              "rules": Array [
                                "transition:background-color ",
                                "100ms",
                                " ease-out,border-color ",
                                "100ms",
                                " ease-out,box-shadow ",
                                "100ms",
                                " ease-out,color ",
                                "100ms",
                                " ease-out;",
                                [Function],
                              ],
                            },
                            "displayName": "CheckboxIndicator",
                            "foldedComponentIds": Array [],
                            "render": [Function],
                            "styledComponentId": "CheckboxIndicator-sc-16t4mhx-0",
                            "target": "span",
                            "toString": [Function],
                            "warnTooManyClasses": [Function],
                            "withComponent": [Function],
                          }
                        }
                        forwardedRef={null}
                      >
                        <span
                          className="c4"
                        />
                      </StyledComponent>
                    </CheckboxIndicator>
                  </span>
                </StyledComponent>
              </ChoiceWrappers__ChoiceInnerWrapper>
            </span>
          </StyledComponent>
        </ChoiceWrappers__ChoiceWrapper>
        <ChoiceText>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "ChoiceText-sc-1qnriny-0",
                  "isStatic": false,
                  "lastClassName": "c5",
                  "rules": Array [
                    "margin-left:10px;",
                  ],
                },
                "displayName": "ChoiceText",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "ChoiceText-sc-1qnriny-0",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <span
              className="c5"
            />
          </StyledComponent>
        </ChoiceText>
      </label>
    </StyledComponent>
  </ChoiceLabel>
</Checkbox>
`;
